### PR TITLE
Flink: Enhance DeleteFilesProcessor with Time Metrics and Unit Tests.

### PR DESCRIPTION
**Pull request by slfan1989**
## Summary:

This pull request enhances the DeleteFilesProcessor by adding time metrics for file deletion operations. Additionally, the unit tests for DeleteFilesProcessor are extended with several edge cases and failure scenarios, ensuring the processor works correctly under various conditions, including file deletion failures, large files, batch deletions, and concurrent deletions.

### Added Time Metrics:

Introduced time metrics to DeleteFilesProcessor to track the time taken for each file deletion operation, helping monitor the performance of delete operations.

The Histogram is used to measure and log the time taken for file deletion operations, providing detailed performance data.

### Extended Unit Tests:

- Delete Non-Existent File: Tests the scenario where a file does not exist, ensuring the system handles this case without throwing exceptions.

- Delete Large File: Simulates the deletion of a large file, ensuring the processor can handle large file deletions correctly and performs well.

- Batch Delete Boundary Test: Tests batch file deletion to ensure files are deleted in batches once the batch size is reached.

- Concurrent Delete: Simulates concurrent file deletions, ensuring stability when handling concurrent delete operations.

- Delete With Failure: Simulates file deletion failures, ensuring the processor can handle failures and correctly record the number of failed deletions.

### CODE REVIEW

File: flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DeleteFilesProcessor.java [hunk 1]
```diff
@@ -76,6 +79,10 @@ public void open() throws Exception {
         taskMetricGroup.counter(TableMaintenanceMetrics.DELETE_FILE_FAILED_COUNTER);
     this.succeededCounter =
         taskMetricGroup.counter(TableMaintenanceMetrics.DELETE_FILE_SUCCEEDED_COUNTER);
+    this.deleteFileTimeMsHistogram =
+        taskMetricGroup.histogram(
+            TableMaintenanceMetrics.DELETE_FILE_TIME_MS_HISTOGRAM,
+            new DescriptiveStatisticsHistogram(1000));
```
Comment (pvary) [inline flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DeleteFilesProcessor.java]: @mxm, @Guosmilesmile: Is this a useful metrics based on your experience too?

File: flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DeleteFilesProcessor.java [hunk 2]
```diff
@@ -116,6 +124,23 @@ private void deleteFiles() {
           e);
       succeededCounter.inc(deletedFilesCount);
       failedCounter.inc(e.numberFailedObjects());
+    } finally {
+      long elapsed = System.currentTimeMillis() - startTime;
+      if (deleteFileTimeMsHistogram != null) {
```
Comment (pvary) [inline flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DeleteFilesProcessor.java]: Is this check needed?

File: flink/v2.0/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDeleteFilesProcessor.java [hunk 3]
```diff
@@ -74,27 +79,172 @@ void testDeleteMissingFile() throws Exception {
...
+  void testDeleteNonExistentFile() throws Exception {
+    String nonexistentFile = "nonexistentFile.txt";
+
+    deleteFile(tableLoader(), nonexistentFile, true);
+
+    assertThat(listFiles(table)).isEqualTo(TABLE_FILES);
+  }
+
+  @Test
+  void testDeleteLargeFile() throws Exception {
```
Comment (pvary) [inline flink/v2.0/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDeleteFilesProcessor.java]: I don't understand this test case?
Why would deleting a large file effect anything?

File: flink/v2.0/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDeleteFilesProcessor.java [hunk 4]
```diff
@@ -74,27 +79,172 @@ void testDeleteMissingFile() throws Exception {
...
+
+    // Use the DeleteFilesProcessor to delete the large file
+    deleteFile(tableLoader(), largeFile.toString(), true);
+
+    // Verify that the large file has been deleted
+    files = listFiles(table);
+    assertThat(files).doesNotContain(largeFileName);
+  }
+
+  private void deleteFile(TableLoader tableLoader, String fileName, boolean expectSuccess)
```
Comment (pvary) [inline flink/v2.0/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDeleteFilesProcessor.java]: We usually try to avoid having boolean parameters, because it is hard to read the code without visiting the method declaration and explicitly checking the parameter.
We sometimes accept this in test code, but please leave a comment where they are called. Something like this:
```
deleteFile(tableLoader(), nonexistentFile, true /* expectSuccess */);
```

File: flink/v2.0/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDeleteFilesProcessor.java [hunk 5]
```diff
@@ -74,27 +79,172 @@ void testDeleteMissingFile() throws Exception {
...
+      throws Exception {
```
Comment (pvary) [inline flink/v2.0/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDeleteFilesProcessor.java]: Can we put this private method at the end of the file?

File: flink/v2.0/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDeleteFilesProcessor.java [hunk 6]
```diff
@@ -74,27 +79,172 @@ void testDeleteMissingFile() throws Exception {
...
+      assertThat(deleteFilesProcessor.getFailedCounter().getCount()).isEqualTo(0);
+      assertThat(deleteFilesProcessor.getDeleteFileTimeMsHistogram().getStatistics().getMean())
+          .isGreaterThan(0);
+    } finally {
+      deleteFilesProcessor.close();
+    }
+  }
+
+  @Test
+  void testConcurrentDelete() throws Exception {
```
Comment (pvary) [inline flink/v2.0/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDeleteFilesProcessor.java:132]: How often this delete actually happens concurrently?
Like one file deleted by the `DeleteFilesProcessor`, and one file deleted by the `executorService`?

If we want to test real concurrency, we likely need to use more files, and maybe some delays.
If we are ok with checking concurrently deleted files, then `testDeleteNonExistentFile` might be enough

File: flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DeleteFilesProcessor.java (hunk 1 above)
Comment (Guosmilesmile) [reply flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DeleteFilesProcessor.java]: In my use case, I haven't focused on the time it takes to delete files. The only time I've encountered long deletion times is when IO is high, and in those cases, write performance or machine metrics would directly reflect the issue. @slfan1989 Could you describe the specific scenario you're dealing with?

Comment (slfan1989) [reply flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DeleteFilesProcessor.java]: @pvary @Guosmilesmile We are currently investigating the integration of the ODS layer with Iceberg, though we do not yet have a concrete business scenario. Nevertheless, I believe adding a metric for the duration of delete operations is valuable: it introduces very low overhead but can significantly enhance system observability, helping us identify potential bottlenecks and analyze performance trends. For example, if delete times increase abnormally, we can further investigate whether the issue stems from the underlying object storage/HDFS performance or slower HMS responses.

In this sense, latency histograms provide a more sensitive signal than simply tracking success/failure counts. Even if deletes succeed, a spike in latency can quickly reveal hidden storage or service bottlenecks, which would otherwise be difficult to detect.

Comment (slfan1989) [reply flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DeleteFilesProcessor.java]: I really appreciate your valuable input, @pvary. I’ll take some time to reflect on your suggestions and share a considered reply soon.

File: flink/v2.0/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDeleteFilesProcessor.java (hunk 3 above)
Comment (slfan1989) [reply flink/v2.0/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDeleteFilesProcessor.java]: @pvary This is a very good question! In most practical scenarios, the metadata files of Iceberg tables are usually not very large. However, considering our internal use cases, I wrote this unit test to validate potential risks.

We plan to provide feature data for the company’s internal AI team. According to user feedback, these feature tables may contain hundreds of columns and require support for multiple concurrent writing tasks. This combination of high dimensionality and high concurrency could cause metadata files to grow rapidly.

Therefore, we need to strike a balance between two strategies:

- More, smaller metadata files — beneficial for concurrent writes but increases metadata management overhead.
- Fewer, larger metadata files — reduces the number of files but may increase latency for individual operations.

Currently, the default value of MANIFEST_TARGET_SIZE_BYTES in our system is 8 MB:

```
public static final String MANIFEST_TARGET_SIZE_BYTES = "commit.manifest.target-size-bytes";
public static final long MANIFEST_TARGET_SIZE_BYTES_DEFAULT = 8 * 1024 * 1024; // 8 MB
```

We are now evaluating the appropriate value for this parameter. One suggestion is to increase it to 256 MB. Next, we will conduct performance tests to compare throughput and latency under different configurations, so that we can determine the optimal setting.

 I’d really like to hear about your experiences with metadata management. If you could share some practical approaches and insights, that would be fantastic.

cc: @mxm @Guosmilesmile

File: flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DeleteFilesProcessor.java (hunk 2 above)
Comment (slfan1989) [reply flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DeleteFilesProcessor.java]: You're right, this check isn’t necessary here. I’ll improve this part of the code.

File: flink/v2.0/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDeleteFilesProcessor.java (hunk 4 above)
Comment (slfan1989) [reply flink/v2.0/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDeleteFilesProcessor.java]: Thanks for the suggestion! I’ll add comments to this function and mark the parameters at the call sites the way you recommended.

File: flink/v2.0/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDeleteFilesProcessor.java (hunk 6 above)
Comment (slfan1989) [reply flink/v2.0/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDeleteFilesProcessor.java:132]: > How often this delete actually happens concurrently?
> Like one file deleted by the `DeleteFilesProcessor`, and one file deleted by the `executorService`?

You are right — the way I previously simulated concurrency may not reflect a realistic scenario. Instead, what we really want to validate is the case where multiple tasks operate on the same table at the same time, leading to concurrent deletions. To cover this, I have improved the unit test by using two DeleteFilesProcessor instances deleting files concurrently.

> If we want to test real concurrency, we likely need to use more files, and maybe some delays.
> If we are ok with checking concurrently deleted files, then `testDeleteNonExistentFile` might be enough

In the latest test, I ensured that both processors attempt to delete the same files concurrently, which allows us to verify that DeleteFilesProcessor behaves as expected under concurrent conditions.

From my local run, I can observe both processors deleting files at the same time:

```
[deleter-p2] INFO org.apache.iceberg.flink.maintenance.operator.DeleteFilesProcessor - Deleted 2 files from table hadoop.default.t using bulk deletes
[deleter-p1] INFO org.apache.iceberg.flink.maintenance.operator.DeleteFilesProcessor - Deleted 2 files from table hadoop.default.t using bulk deletes
[deleter-p1] INFO org.apache.iceberg.flink.maintenance.operator.DeleteFilesProcessor - Deleted 2 files from table hadoop.default.t using bulk deletes
[deleter-p2] INFO org.apache.iceberg.flink.maintenance.operator.DeleteFilesProcessor - Deleted 2 files from table hadoop.default.t using bulk deletes
[deleter-p2] INFO org.apache.iceberg.flink.maintenance.operator.DeleteFilesProcessor - Deleted 2 files from table hadoop.default.t using bulk deletes
```

File: flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DeleteFilesProcessor.java [hunk 7]
```diff
@@ -116,6 +124,21 @@ private void deleteFiles() {
...
+    return failedCounter;
+  }
+
+  public Counter getSucceededCounter() {
+    return succeededCounter;
+  }
+
+  public Histogram getDeleteFileTimeMsHistogram() {
+    return deleteFileTimeMsHistogram;
+  }
```
Comment (pvary) [inline flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DeleteFilesProcessor.java]: Please remove the `get` from the method name. See: https://iceberg.apache.org/contribute/#method-naming
It might also worth to read through the other style guides as well.

Also these methods should not be public. They should be package private and should be annotated with `@VisibleForTesting` annotation. Check the other cases in the Flink code

File: flink/v2.0/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDeleteFilesProcessor.java [hunk 8]
```diff
@@ -74,27 +79,183 @@ void testDeleteMissingFile() throws Exception {
...
+
+    // Generate 30 test files: delete-0.txt ... delete-29.txt
+    Set<String> targets = Sets.newHashSet();
+    for (int i = 0; i < 30; i++) {
+      targets.add("delete-" + i + ".txt");
+    }
+
+    for (String f : targets) {
+      Files.write(root.resolve(f), f.getBytes(StandardCharsets.UTF_8));
+    }
```
Comment (pvary) [inline flink/v2.0/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDeleteFilesProcessor.java]: Why not merge the 2 `for` loops?

Also add newline after every block. See: https://iceberg.apache.org/contribute/#block-spacing

File: flink/v2.0/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDeleteFilesProcessor.java [hunk 9]
```diff
@@ -74,27 +79,183 @@ void testDeleteMissingFile() throws Exception {
...
+                  for (String f : targets) {
+                    barriers.get(f).await(2, TimeUnit.SECONDS);
+                    h1.processElement(f, ts);
+                  }
+                  h1.processWatermark(EVENT_TIME);
+                  h1.endInput();
+                } catch (Exception ignored) {
+                }
+              },
+              "deleter-p1");
```
Comment (pvary) [inline flink/v2.0/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDeleteFilesProcessor.java]: Do we have to do these in threads? Could we just advance the processors in a loop?

File: flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DeleteFilesProcessor.java (hunk 7 above)
Comment (slfan1989) [reply flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DeleteFilesProcessor.java]: Thank you for reviewing the code. I will make improvements based on your suggestions as soon as possible.

Comment (slfan1989) [reply flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DeleteFilesProcessor.java]: > Please remove the get from the method name. See: https://iceberg.apache.org/contribute/#method-naming
> It might also worth to read through the other style guides as well.

> Also these methods should not be public. They should be package private and should be annotated with 
> @VisibleForTesting annotation. Check the other cases in the Flink code

Thank you for the information! I have improved the code based on your suggestions.

File: flink/v2.0/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDeleteFilesProcessor.java (hunk 8 above)
Comment (slfan1989) [reply flink/v2.0/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDeleteFilesProcessor.java]: I have made improvements to this part of the code.

File: flink/v2.0/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDeleteFilesProcessor.java (hunk 9 above)
Comment (slfan1989) [reply flink/v2.0/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDeleteFilesProcessor.java]: Thank you for your question; it's a great one. The purpose of designing this unit test was to simulate a scenario where two Flink tasks are concurrently performing expiration snapshot cleanup, which leads to the concurrent deletion of metadata files. To model this, I used threads. After carefully considering your suggestion, I realized that using a loop to sequentially process both processors for deletion is also a valid approach. I've since made the necessary improvements to the code.

**Comment by slfan1989** [comment]
@pvary Could you please take a look at this PR? Thank you very much!

File: flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DeleteFilesProcessor.java (hunk 1 above)
Comment (mxm) [reply flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DeleteFilesProcessor.java]: We haven't required the need to measure delete file duration. We are generally more concerned with the total time the maintenance takes. I would assume that the actual deletion of the files is a large chunk, but we have managed storage, so there is usually very little we can do about it. Maybe it would suffice to keep the max duration here instead of a full histogram?

File: flink/v2.0/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDeleteFilesProcessor.java [hunk 10]
```diff
@@ -74,27 +75,158 @@ void testDeleteMissingFile() throws Exception {
...
+    Set<String> files = listFiles(table);
+    assertThat(files).contains(largeFileName);
+
+    // Use the DeleteFilesProcessor to delete the large file
+    deleteFile(tableLoader(), largeFile.toString(), true /* expectSuccess */);
+
+    // Verify that the large file has been deleted
+    files = listFiles(table);
+    assertThat(files).doesNotContain(largeFileName);
+  }
```
Comment (mxm) [inline flink/v2.0/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDeleteFilesProcessor.java]: I'm sure that we can delete a 10Mb file, at least on local file system. I'm not sure what we are testing here. Why does the size matter here?

File: flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DeleteFilesProcessor.java (hunk 1 above)
Comment (slfan1989) [reply flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/DeleteFilesProcessor.java]: @mxm Thank you very much for your suggestion! I have made the changes to the code based on your suggestions. Could you please take another look?

File: flink/v2.0/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDeleteFilesProcessor.java (hunk 10 above)
Comment (slfan1989) [reply flink/v2.0/flink/src/test/java/org/apache/iceberg/flink/maintenance/operator/TestDeleteFilesProcessor.java]: Apologies for the inconvenience. Since I am responsible for the development of part of the storage software at the company, I’ve been using a fixed testing approach (such as deleting large files), but these methods may not be fully compatible with the current scenario. I agree with your point that this unit test is unnecessary, and I have already removed it.
